src/System.Security.Cryptography.Pkcs/src/Internal/Cryptography/Pal/Windows/DecryptorPalWindows.Decrypt.cs

```diff
@@ -0,0 +1,223 @@
+            if (exception != null)
+                return null;
+
+            // Desktop compat: We pass false for "silent" here (thus allowing crypto providers to display UI.)
+            using (SafeProvOrNCryptKeyHandle hKey = TryGetCertificatePrivateKey(cert, false, out exception))
+            {
+                if (hKey == null)
+                    return null;
+
+                RecipientInfoType type = recipientInfo.Type;
+                switch (type)
+                {
+                    case RecipientInfoType.KeyTransport:
+                        exception = TryDecryptTrans((KeyTransRecipientInfo)recipientInfo, hKey, keySpec);
+                        break;
+
+                    case RecipientInfoType.KeyAgreement:
+                        exception = TryDecryptAgree((KeyAgreeRecipientInfo)recipientInfo, hKey, keySpec, originatorCerts, extraStore);
+                        break;
+
+                    default:
+                        // Since only the framework can construct RecipientInfo's, we're at fault if we get here. So it's okay to assert and throw rather than 
+                        // returning to the caller.
+                        Debug.Fail($"Unexpected RecipientInfoType: {type}");
+                        throw new NotSupportedException();
+                }
+
+                if (exception != null)
+                    return null;
+
+                // If we got here, we successfully decrypted. Return the decrypted content.
+                return _hCryptMsg.GetContentInfo();
+            }
+        }
+
+        private static Exception TryGetKeySpecForCertificate(X509Certificate2 cert, out CryptKeySpec keySpec)
+        {
+            using (SafeCertContextHandle hCertContext = cert.CreateCertContextHandle())
+            {
+                int cbSize = 0;
+                if (!Interop.Crypt32.CertGetCertificateContextProperty(hCertContext, CertContextPropId.CERT_KEY_PROV_INFO_PROP_ID, null, ref cbSize))
+                {
+                    ErrorCode errorCode = (ErrorCode)(Marshal.GetLastWin32Error());
+                    keySpec = default(CryptKeySpec);
+                    return errorCode.ToCryptographicException();
+                }
+
+                byte[] pData = new byte[cbSize];
+                unsafe
+                {
+                    fixed (byte* pvData = pData)
+                    {
+                        if (!Interop.Crypt32.CertGetCertificateContextProperty(hCertContext, CertContextPropId.CERT_KEY_PROV_INFO_PROP_ID, pData, ref cbSize))
+                        {
+                            ErrorCode errorCode = (ErrorCode)(Marshal.GetLastWin32Error());
+                            keySpec = default(CryptKeySpec);
+                            return errorCode.ToCryptographicException();
+                        }
+
+                        CRYPT_KEY_PROV_INFO* pCryptKeyProvInfo = (CRYPT_KEY_PROV_INFO*)pvData;
+                        keySpec = pCryptKeyProvInfo->dwKeySpec;
```

Comment:
On the one hand, I'm curious what the value of this is for a CNG key (-1? 0?).  On the other, the field it gets passed into says it's ignored when hKey is an NCrypt handle; so that's entirely an academic wonder on my part.